### ISSUE TITLE
Scaling ( dividing the raw 32 bit integer )

### ISSUE THREAD

**Issue by timCF** [jj1bdx/sfmt-erlang#23]
Even after the float range bug has been fixed to return numbers between 0 and 1 (non-inclusive), the modulo bias will still remain because the float value computed by dividing the raw 32 bit integer RNG output by 2^32 will still have exactly 2^32 discrete values. The scaling performed by multiplying with the range just maps these 2^32 discrete values into N scaled numbers where N is the range. In cases where N is not  an exact power of 2, this will result in a situation where the frequency of occurrence of every value in the scaled range can never be the same mathematically. (The generalized version of the Pigeonhole principle :  https://en.wikipedia.org/wiki/Pigeonhole_principle ). The only way in which this modulo bias can be avoided is by throwing away some raw numbers at the higher end of the range

**Comment by jj1bdx** [comment]
@timCF So do you want something like this, which I once used for sfmt-erlang too (see past source code)? https://github.com/jj1bdx/tinymt-erlang/blob/master/src/tinymt32.erl#L403

**Comment by timCF** [comment]
@jj1bdx 

In “sfmt,erl” line 438 – uniform() is computed by calling the “gen_rand_float” function – This is performed in the C code by dividing the raw 32 bit number by  2^32 (floating point division yielding a double result between 0.0 and 1.0).
 
Line 457/458 of “sfmt.erl” is the one where the scaling is performed – it is done using the construct :  “trunc(uniform() * N) + 1”. So this returns values in the range 1 to N.  (The same is done in lines 476 to 478 of “sfmt.erl” also.
 
Since  a 32 bit number raw number can have only 2^32 (call this M) discrete values, the double returned by uniform() or gen_rand_float() will also have only “M” discrete values – it is not continuously variable. When you perform the scaling by “trunc(uniform() * N)”, each of these M steps will get translated to a number between 0 and N-1 (N discrete values). So essentially you're mapping M possible values of the unscaled number to N distinct values in the scaled range.
 
In order for scaling not to introduce any extra bias, the distribution of the output should be such that each possible value in the scaled range N should occur with exactly the same probability. But in cases where M is not exactly divisible by N this is mathematically impossible because the following situation occurs.
 
Assume M/N (integer division) = k with a remainder of r
 
In the output after scaling, “r” numbers will occur (k+1) times, and (N-r) numbers can occur only “k” times (The pigeonhole principle). This is what constitutes the bias. The scaling will be completely bias-free only in cases where M is exactly divisible by N, that is when N is a power of 2.
 
The proper way to perform bias free scaling :
 
a) Compute the maximum value of the 32 bit number which is a multiple of N, i.e (k*N).
b) When the raw number returned is equal to or more than this number (there will be exactly “r” instances in which this occurs), throw away that number, and draw another number afresh.
c) Now you can perform the scaling by a simple modulo operator. (Even a floating point multiply will yield uniformly distributed values, but integer modulo operator is generally more efficient).
 
Since you're discarding all numbers above k*N, the range of raw numbers you're using now is an exact multiple of N – and so its distribution after the scaling operation will also be perfectly uniform.

**Comment by jj1bdx** [comment]
@timCF I understand and I know what you are talking about.

I abandoned the code for compatibility behavior to the random and rand modules.  If you dig into the history of sfmt-erlang archive, you would find out when I abandoned the code. I even implemented the stuff in the NIFs, so what I should do to meet your needs is simply rolling back the code for giving equal distribution for all possible N outputs, where the single output X is 1 <= X <= N.

The following code is from tinymt-erlang, which I keep retaining the pideonhole principle abiding code. (next_state/1 and temper/1 are the pair operation which makes the RNG one step further, FYI.)

Prof. Makoto Matsumoto has pointed out the same issue. See http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/efaq.html at the section *Want a uniform random number in the integers [0..N-1]*.

```erlang
uniform_s(Max, R) when is_integer(Max), Max >= 1 ->
    Limit = ?TWOPOW32 - (?TWOPOW32 rem Max),
    uniform_s(Max, Limit, R).

uniform_s(M, L, R) ->
    R1 = next_state(R),
    V = temper(R1),
    case V < L of
    true -> {(V rem M) + 1, R1};
    false -> uniform_s(M, L, R1)
    end.
```

**Comment by jj1bdx** [comment]
s/rolling back/resurrecting (or reviving)/

**Comment by jj1bdx** [comment]
@timCF 

The issue I was worrying about literally implementing the pigeonhole principle is that the operation might cause non-deterministic number of multiple retries; for example, if the source random number sequence is something like [10, 11, 12, 13, 12, 11, 10, 11, 12, ...] and you need to pick up from the numbers where one of the numbers X is 0 <= X <= 9, the function will keep looping forever until the condition of X is satisfied. **This sort of extreme condition may highly unlikely to happen, I guess, but at the same time I cannot give you a mathematical proof for why the extreme condition will not continue forever, either.**

Maybe I should document the above limitation and revert back the code as a new version. Allow me some time.

**Comment by jj1bdx** [comment]
@timCF

The probability of having to discard the 32-bit raw output R for [1, N] range integer output is `(N - (2^32 rem N)) / (2^32)`, which is very small if N is small.

**Comment by timCF** [comment]
> Maybe I should document the above limitation and revert back the code as a new version. Allow me some time.

Ok, thx for consideration of issue!

**Comment by jj1bdx** [comment]
See #24

TODO:
* Feedback the code to sfmt_pure.erl
* Rewrite the NIF code
* Write new integer tests
* Milestone: 0.13.0

**Comment by jj1bdx** [comment]
Scaling issue fix code incorporated into 0.13.0. Closed.
